index.js

```diff
@@ -1,53 +1,57 @@
 
-export default multiformats => {
-  const { CID, bytes, multibase } = multiformats
-  const _encode = obj => transform(obj, (result, value, key) => {
-    const cid = CID.asCID(value)
-    if (cid) {
-      result[key] = { '/': cid.toString() }
-    } else if (bytes.isBinary(value)) {
-      value = bytes.coerce(value)
-      result[key] = { '/': { bytes: multibase.encode(value, 'base64') } }
-    } else if (typeof value === 'object' && value !== null) {
-      result[key] = _encode(value)
-    } else {
-      result[key] = value
-    }
-  })
+const _encode = (obj) => transform(obj, (result, value, key) => {
+  const cid = CID.asCID(value)
+  if (cid) {
+    result[key] = { '/': cid.toString() }
+  } else if (bytes.isBinary(value)) {
+    value = bytes.coerce(value)
+    result[key] = { '/': { bytes: base64.encode(value) } }
+  } else if (typeof value === 'object' && value !== null) {
+    result[key] = _encode(value)
+  } else {
+    result[key] = value
+  }
+})
 
-  const encode = obj => {
-    if (typeof obj === 'object' && !bytes.isBinary(obj) && !CID.asCID(obj) && obj) {
-      if (isCircular(CID, obj)) throw new Error('Object contains circular references.')
-      obj = _encode(obj)
+const encode = (obj) => {
+  if (typeof obj === 'object' && !bytes.isBinary(obj) && !CID.asCID(obj) && obj) {
+    if (isCircular(obj, { asCID: true })) {
+      throw new Error('Object contains circular references')
     }
-    return bytes.fromString(json(obj))
+    obj = _encode(obj)
   }
+  return bytes.fromString(json(obj))
+}
 
-  const _decode = obj => transform(obj, (result, value, key) => {
-    if (typeof value === 'object' && value !== null) {
-      if (value['/']) {
-        if (typeof value['/'] === 'string') {
-          result[key] = CID.from(value['/'])
-        } else if (typeof value['/'] === 'object' && value['/'].bytes) {
-          result[key] = multibase.decode(value['/'].bytes, 'base64')
-        } else {
-          result[key] = _decode(value)
-        }
+const _decode = (obj) => transform(obj, (result, value, key) => {
+  if (typeof value === 'object' && value !== null) {
+    if (value['/']) {
+      if (typeof value['/'] === 'string') {
+        result[key] = CID.parse(value['/'])
+      } else if (typeof value['/'] === 'object' && value['/'].bytes) {
+        result[key] = base64.decode(value['/'].bytes)
       } else {
         result[key] = _decode(value)
       }
     } else {
-      result[key] = value
+      result[key] = _decode(value)
     }
-  })
-  const decode = buffer => {
-    const obj = JSON.parse(bytes.toString(buffer))
-    return _decode({ value: obj }).value
+  } else {
+    result[key] = value
   }
-  const name = 'dag-json'
-  const code = 0x0129
-  return { encode, decode, name, code }
+})
+
+const decode = (buffer) => {
+  const obj = JSON.parse(bytes.toString(buffer))
+  return _decode({ value: obj }).value
 }
+
+const name = 'dag-json'
+const code = 0x0129
+
+export { encode, decode, name, code }
```

Comment:
I would suggest following instead:

```js
import { codec } from 'multiformats/codecs/codec'
export default codec({ name, code, encode, decode })
```

Because it will:

1. Create a codec that is implements both encoder and decoder ends
2. Expose `encoder` and `decoder` ends separately.
3. If we do end up adding some kind of contaminators, `dag-json` would automatically get that as well.